code/game/objects/items/devices/laserpointer.dm

```diff
@@ -44,128 +64,226 @@
 		return TRUE
 
-/obj/item/laser_pointer/attackby(obj/item/W, mob/user, params)
-	if(istype(W, /obj/item/stock_parts/micro_laser))
-		if(!diode)
-			if(!user.transferItemToLoc(W, src))
-				return
-			diode = W
-			to_chat(user, span_notice("You install a [diode.name] in [src]."))
-		else
-			to_chat(user, span_warning("[src] already has a diode installed!"))
-	else
-		return ..()
+/obj/item/laser_pointer/tool_act(mob/living/user, obj/item/tool)
+	. = ..()
+	if(isnull(crystal_lens) || !(tool.tool_behaviour == TOOL_WIRECUTTER || tool.tool_behaviour == TOOL_HEMOSTAT))
+		return
+	tool.play_tool_sound(src)
+	balloon_alert(user, "removed crystal lens")
+	crystal_lens.forceMove(drop_location())
+	crystal_lens = null
+	return TRUE
+
+/obj/item/laser_pointer/attackby(obj/item/attack_item, mob/user, params)
+	if(istype(attack_item, /obj/item/stock_parts/micro_laser))
+		if(diode)
+			balloon_alert(user, "already has a diode!")
+			return
+		var/obj/item/stock_parts/attack_diode = attack_item
+		if(crystal_lens && attack_diode.rating < 3) //only tier 3 and up are small enough to fit
+			to_chat(user, span_warning("You try to jam \the [attack_item.name] in place, but \the [crystal_lens.name] is in the way!"))
+			playsound(src, 'sound/machines/airlock_alien_prying.ogg', 20)
+			if(do_after(user, 2 SECONDS, src))
+				var/atom/atom_to_teleport = pick(user, attack_item)
+				if(atom_to_teleport == user)
+					to_chat(user, span_warning("You jam \the [attack_item.name] in too hard and break \the [crystal_lens.name] inside, teleporting you away!"))
+					user.drop_all_held_items()
+				else if(atom_to_teleport == attack_item)
+					attack_item.forceMove(drop_location())
+					to_chat(user, span_warning("You jam \the [attack_item.name] in too hard and break \the [crystal_lens.name] inside, teleporting \the [attack_item.name] away!"))
+				do_teleport(atom_to_teleport, get_turf(src), crystal_lens.blink_range, asoundin = 'sound/effects/phasein.ogg', channel = TELEPORT_CHANNEL_BLUESPACE)
+				qdel(crystal_lens)
+			return
+		if(!user.transferItemToLoc(attack_item, src))
+			return
+		playsound(src, 'sound/items/screwdriver.ogg', 30)
+		diode = attack_item
+		balloon_alert(user, "installed \the [diode.name]")
+		//we have a diode now, try starting a charge sequence in case the pointer was charging when we took out the diode
+		recharging = TRUE
+		START_PROCESSING(SSobj, src)
+		return TRUE
+
+	if(istype(attack_item, /obj/item/stack/ore/bluespace_crystal))
+		if(crystal_lens)
+			balloon_alert(user, "already has a lens!")
+			return
+		//the crystal stack we're trying to install a crystal from
+		var/obj/item/stack/ore/bluespace_crystal/crystal_stack = attack_item
+		if(diode && diode.rating < 3) //only lasers of tier 3 and up can house a lens
+			to_chat(user, span_warning("You try to jam \the [crystal_stack.name] in front of the diode, but it's a bad fit!"))
+			playsound(src, 'sound/machines/airlock_alien_prying.ogg', 20)
+			if(do_after(user, 2 SECONDS, src))
```

Comment:
I think these do_after actions and what follows could share a common proc.

Reply:
i'd agree if i made them equal in possible results which i didn't nor do i intend to. one's teleporting either the diode or the user, the other's teleporting either the pointer or the user; writing out more typechecks for differing `to_chat()` messages would probably just make it more of a mess than it should be